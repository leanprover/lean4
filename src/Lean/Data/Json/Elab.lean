/-
Copyright (c) 2022 E.W.Ayers. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: E.W.Ayers, Wojciech Nawrocki
-/
module

prelude
public import Lean.Data.Json.FromToJson
public meta import Lean.Syntax

public section

/-!
# JSON-like syntax for Lean.

Now you can write

```lean
open Lean.Json

#eval json% {
  hello : "world",
  cheese : ["edam", "cheddar", {kind : "spicy", rank : 100.2}],
  lemonCount : 100e30,
  isCool : true,
  isBug : null,
  lookACalc: $(23 + 54 * 2)
}
```
-/

namespace Lean.Json

/-- Json syntactic category -/
declare_syntax_cat json (behavior := symbol)
/-- Json null value syntax. -/
syntax "null" : json
/-- Json true value syntax. -/
syntax "true" : json
/-- Json false value syntax. -/
syntax "false" : json
/-- Json string syntax. -/
syntax str : json
/-- Json number negation syntax for ordinary numbers. -/
syntax "-"? num : json
/-- Json number negation syntax for scientific numbers. -/
syntax "-"? scientific : json
/-- Json array syntax. -/
syntax "[" json,* "]" : json
/-- Json identifier syntax. -/
syntax jsonIdent := ident <|> str
/-- Json key/value syntax. -/
syntax jsonField := jsonIdent ": " json
/-- Json object syntax. -/
syntax "{" jsonField,* "}" : json
/-- Allows to use Json syntax in a Lean file. -/
syntax "json% " json : term


macro_rules
  | `(json% null)           => `(Lean.Json.null)
  | `(json% true)           => `(Lean.Json.bool Bool.true)
  | `(json% false)          => `(Lean.Json.bool Bool.false)
  | `(json% $n:str)         => `(Lean.Json.str $n)
  | `(json% $n:num)         => `(Lean.Json.num $n)
  | `(json% $n:scientific)  => `(Lean.Json.num $n)
  | `(json% -$n:num)        => `(Lean.Json.num (-$n))
  | `(json% -$n:scientific) => `(Lean.Json.num (-$n))
  | `(json% [$[$xs],*])     => `(Lean.Json.arr #[$[json% $xs],*])
  | `(json% {$[$ks:jsonIdent : $vs:json],*}) => do
    let ks : Array (TSyntax `term) ← ks.mapM fun
      | `(jsonIdent| $k:ident) => pure (k.getId |> toString |> quote)
      | `(jsonIdent| $k:str)   => pure k
      | _                     => Macro.throwUnsupported
    `(Lean.Json.mkObj [$[($ks, json% $vs)],*])
  | `(json% $stx)           =>
    if stx.raw.isAntiquot then
      let stx := ⟨stx.raw.getAntiquotTerm⟩
      `(Lean.toJson $stx)
    else
      Macro.throwUnsupported

end Lean.Json
