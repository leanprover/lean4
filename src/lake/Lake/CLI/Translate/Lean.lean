/-
Copyright (c) 2024 Mac Malone. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Mac Malone
-/
prelude
import Lake.DSL
import Lake.Config.Package
import Lean.Parser.Module

/-! # Lean Translation

Converts a declarative Lake configuration into a Lean module.
-/

namespace Lake
open DSL System Lean Syntax Parser Module

/-! ## General Helpers -/

private local instance : Quote FilePath where
  quote path := quote path.toString

private local instance [Quote α] : Quote (Array α) where
  quote xs := let xs : Array Term := xs.map quote; Unhygienic.run `(#[$xs,*])

private local instance : BEq FilePath where
  beq a b := a.normalize == b.normalize

private local instance : Quote Bool where
  quote b := mkIdent <| if b then `true else `false

@[inline] def addDeclField [Quote α] (name : Name) (val : α) (fs : Array DeclField) : Array DeclField :=
  fs.push <| Unhygienic.run `(declField|$(mkIdent name) := $(quote val))

@[inline] def addDeclField? [Quote α] (name : Name) (val? : Option α) (fs : Array DeclField) : Array DeclField :=
  if let some val := val? then addDeclField name val fs else fs

@[inline] def addDeclFieldD [Quote α] [BEq α] (name : Name) (val : α) (default : α) (fs : Array DeclField) : Array DeclField :=
  if val == default then fs else addDeclField name val fs

@[inline] def addDeclFieldNotEmpty [Quote α] (name : Name) (val : Array α) (fs : Array DeclField) : Array DeclField :=
  if val.isEmpty then fs else addDeclField name val fs

/-! ## Value Encoders -/

protected def BuildType.quote : BuildType → Term
| .debug => mkCIdent ``debug
| .minSizeRel => mkCIdent ``minSizeRel
| .relWithDebInfo => mkCIdent ``relWithDebInfo
| .release => mkCIdent ``release

instance : Quote BuildType := ⟨BuildType.quote⟩

protected def Backend.quote : Backend → Term
| .c => mkCIdent ``c
| .llvm => mkCIdent ``llvm
| .default => mkCIdent ``default

instance : Quote Backend := ⟨Backend.quote⟩

def quoteLeanOptionValue : LeanOptionValue → Term
| .ofString v => quote v
| .ofBool v => quote v
| .ofNat v => quote v

private local instance : Quote LeanOptionValue := ⟨quoteLeanOptionValue⟩

def quoteLeanOption (opt : LeanOption) : Term := Unhygienic.run do
  `(⟨$(quote opt.name), $(quote opt.value)⟩)

private local instance : Quote LeanOption := ⟨quoteLeanOption⟩

protected def LeanVer.quote (v : LeanVer) : Term := Unhygienic.run do
  let lit := Syntax.mkLit interpolatedStrLitKind  v.toString.quote
  let stx := mkNode interpolatedStrKind #[lit]
  `(v!$stx)

private local instance : Quote LeanVer := ⟨LeanVer.quote⟩

/-! ## Configuration Encoders -/

def WorkspaceConfig.addDeclFields (cfg : WorkspaceConfig) (fs : Array DeclField) : Array DeclField :=
  addDeclFieldD `packagesDir cfg.packagesDir defaultPackagesDir fs

def LeanConfig.addDeclFields (cfg : LeanConfig) (fs : Array DeclField) : Array DeclField :=
  fs
  |> addDeclFieldD `buildType  cfg.buildType .release
  |> addDeclFieldD `backend cfg.backend .default
  |> addDeclField? `platformIndependent cfg.platformIndependent
  |> addDeclFieldNotEmpty `leanOptions cfg.leanOptions
  |> addDeclFieldNotEmpty `moreServerOptions cfg.moreServerOptions
  |> addDeclFieldNotEmpty `moreLeanArgs cfg.moreLeanArgs
  |> addDeclFieldNotEmpty `weakLeanArgs cfg.weakLeanArgs
  |> addDeclFieldNotEmpty `moreLeancArgs cfg.moreLeancArgs
  |> addDeclFieldNotEmpty `weakLeancArgs cfg.weakLeancArgs
  |> addDeclFieldNotEmpty `moreLinkArgs cfg.moreLinkArgs
  |> addDeclFieldNotEmpty `weakLinkArgs cfg.weakLinkArgs

@[inline] def mkDeclValWhere? (fields : Array DeclField) : Option (TSyntax ``declValWhere) :=
  if fields.isEmpty then none else Unhygienic.run `(declValWhere|where $fields*)

def PackageConfig.mkSyntax (cfg : PackageConfig)
  (testDriver := cfg.testDriver) (lintDriver := cfg.lintDriver)
: PackageDecl := Unhygienic.run do
  have : Quote Term := ⟨id⟩
  let declVal? := mkDeclValWhere? <| Array.empty
    |> addDeclFieldD `precompileModules cfg.precompileModules false
    |> addDeclFieldD `moreGlobalServerArgs cfg.moreGlobalServerArgs #[]
    |> addDeclFieldD `srcDir cfg.srcDir "."
    |> addDeclFieldD `buildDir cfg.buildDir defaultBuildDir
    |> addDeclFieldD `leanLibDir cfg.leanLibDir defaultLeanLibDir
    |> addDeclFieldD `nativeLibDir cfg.nativeLibDir defaultNativeLibDir
    |> addDeclFieldD `binDir cfg.binDir defaultBinDir
    |> addDeclFieldD `irDir cfg.irDir defaultIrDir
    |> addDeclField? `releaseRepo (cfg.releaseRepo <|> cfg.releaseRepo?)
    |> addDeclFieldD `buildArchive (cfg.buildArchive?.getD cfg.buildArchive) (defaultBuildArchive cfg.name)
    |> addDeclFieldD `preferReleaseBuild cfg.preferReleaseBuild false
    |> addDeclFieldD `testDriver testDriver ""
    |> addDeclFieldD `testDriverArgs cfg.testDriverArgs #[]
    |> addDeclFieldD `lintDriver lintDriver ""
    |> addDeclFieldD `lintDriverArgs cfg.lintDriverArgs #[]
    |> addDeclFieldD `version cfg.version {}
    |> addDeclField? `versionTags (quoteVerTags? cfg.versionTags)
    |> addDeclFieldD `description cfg.description ""
    |> addDeclFieldD `keywords cfg.keywords #[]
    |> addDeclFieldD `homepage cfg.homepage ""
    |> addDeclFieldD `license cfg.license ""
    |> addDeclFieldD `licenseFiles cfg.licenseFiles #["LICENSE"]
    |> addDeclFieldD `readmeFile cfg.readmeFile "README.md"
    |> addDeclFieldD `reservoir cfg.reservoir true
    |> cfg.toWorkspaceConfig.addDeclFields
    |> cfg.toLeanConfig.addDeclFields
  `(packageDecl|package $(mkIdent cfg.name):ident $[$declVal?]?)
  where
    quoteVerTags? (pat : StrPat) : Option Term :=
      match pat with
      | .mem xs => if xs.isEmpty then Unhygienic.run `(∅) else some (quote xs)
      | .startsWith pre => Unhygienic.run `(.$(mkIdent `startsWith) $(quote pre))
      | .satisfies _ n =>
        if n.isAnonymous || n == `default then none else
        Unhygienic.run `(.$(mkIdent n))

private def getEscapedNameParts? (acc : List String) : Name → Option (List String)
  | Name.anonymous => if acc.isEmpty then none else some acc
  | Name.str n s => do
    let s ← Name.escapePart s
    getEscapedNameParts? (s::acc) n
  | Name.num _ _ => none

def mkNameLit? (n : Name) : Option NameLit :=
  getEscapedNameParts? [] n  |>.map fun ss => mkNameLit ("`" ++ ".".intercalate ss)

protected def Glob.quote (glob : Glob) : Term := Unhygienic.run do
  match glob with
  | .one n => return quote n
  | .submodules n =>
    match mkNameLit? n with
    | some lit =>`($lit:name.+)
    | none => return mkCApp ``submodules #[quote n]
  | .andSubmodules n =>
    match mkNameLit? n with
    | some lit =>`($lit:name.*)
    | none => return mkCApp ``andSubmodules #[quote n]

local instance : Quote Glob := ⟨Glob.quote⟩

protected def LeanLibConfig.mkSyntax
  (cfg : LeanLibConfig) (defaultTarget := false)
: LeanLibDecl := Unhygienic.run do
  let declVal? := mkDeclValWhere? <| Array.empty
    |> addDeclFieldD `srcDir cfg.srcDir "."
    |> addDeclFieldD `roots cfg.roots #[cfg.name]
    |> addDeclFieldD `globs cfg.globs (cfg.roots.map .one)
    |> addDeclFieldD `libName cfg.libName (cfg.name.toString (escape := false))
    |> addDeclFieldD `precompileModules cfg.precompileModules false
    |> addDeclFieldD `defaultFacets cfg.defaultFacets #[LeanLib.leanArtsFacet]
    |> cfg.toLeanConfig.addDeclFields
  let attrs? ← if defaultTarget then some <$> `(Term.attributes|@[default_target]) else pure none
  `(leanLibDecl|$[$attrs?:attributes]? lean_lib $(mkIdent cfg.name):ident $[$declVal?]?)

protected def LeanExeConfig.mkSyntax
  (cfg : LeanExeConfig) (defaultTarget := false)
: LeanExeDecl := Unhygienic.run do
  let declVal? := mkDeclValWhere? <| Array.empty
    |> addDeclFieldD `srcDir cfg.srcDir "."
    |> addDeclFieldD `root cfg.root cfg.name
    |> addDeclFieldD `exeName cfg.exeName (cfg.name.toStringWithSep "-" (escape := false))
    |> addDeclFieldD `supportInterpreter cfg.supportInterpreter false
    |> cfg.toLeanConfig.addDeclFields
    let attrs? ← if defaultTarget then some <$> `(Term.attributes|@[default_target]) else pure none
  `(leanExeDecl|$[$attrs?:attributes]? lean_exe $(mkIdent cfg.name):ident $[$declVal?]?)

protected def Dependency.mkSyntax (cfg : Dependency) : RequireDecl := Unhygienic.run do
  let src? ← cfg.src?.mapM fun src =>
    match src with
    | .path dir =>
      `(fromSource|$(quote dir):term)
    | .git url rev? subDir? =>
      `(fromSource|git $(quote url) $[@ $(rev?.map quote)]? $[/ $(subDir?.map quote)]?)
  let ver? ←
    if let some ver := cfg.version? then
      if ver.startsWith "git#" then
        some <$> `(verSpec|git $(quote <| ver.drop 4))
      else
        some <$> `(verSpec|$(quote ver):term)
    else
      pure none
  let scope? := if cfg.scope.isEmpty then none else some (quote cfg.scope)
  let opts? := if cfg.opts.isEmpty then none else some <| Unhygienic.run do
    cfg.opts.foldM (init := mkCIdent ``NameMap.empty) fun stx opt val =>
      `($stx |>.insert $(quote opt) $(quote val))
  `(requireDecl|require $[$scope? /]? $(mkIdent cfg.name):ident $[@ $ver?]?
    $[from $src?]? $[with $opts?]?)

/-! ## Root Encoder -/

/-- Create a Lean module that encodes the declarative configuration of the package. -/
def Package.mkLeanConfig (pkg : Package) : TSyntax ``module := Unhygienic.run do
  let defaultTargets := pkg.defaultTargets.foldl NameSet.insert NameSet.empty
  let pkgConfig := pkg.config.mkSyntax pkg.testDriver pkg.lintDriver
  let requires := pkg.depConfigs.map (·.mkSyntax)
  let leanLibs := pkg.leanLibConfigs.toArray.map fun cfg =>
    cfg.mkSyntax (defaultTargets.contains cfg.name)
  let leanExes := pkg.leanExeConfigs.toArray.map fun cfg =>
    cfg.mkSyntax (defaultTargets.contains cfg.name)
  `(module|
  import $(mkIdent `Lake)
  open $(mkIdent `System) $(mkIdent `Lake) $(mkIdent `DSL)
  $pkgConfig:command
  $[$requires:command]*
  $[$leanLibs:command]*
  $[$leanExes:command]*
  )
