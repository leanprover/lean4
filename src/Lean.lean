/-
Copyright (c) 2020 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
-/
import Lean.Data
import Lean.Compiler
import Lean.Environment
import Lean.Modifiers
import Lean.ProjFns
import Lean.Runtime
import Lean.ResolveName
import Lean.Attributes
import Lean.Parser
import Lean.ReducibilityAttrs
import Lean.Elab
import Lean.Class
import Lean.LocalContext
import Lean.MetavarContext
import Lean.AuxRecursor
import Lean.Meta
import Lean.Util
import Lean.Eval
import Lean.Structure
import Lean.PrettyPrinter
import Lean.CoreM
import Lean.InternalExceptionId
import Lean.Server
import Lean.ScopedEnvExtension
import Lean.DocString
import Lean.DeclarationRange
import Lean.LazyInitExtension
import Lean.Widget
