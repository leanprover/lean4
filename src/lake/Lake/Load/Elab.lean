/-
Copyright (c) 2021 Mac Malone. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Mac Malone
-/
import Lean.Elab.Frontend
import Lake.DSL.Extensions
import Lake.DSL.Attributes
import Lake.Load.Config
import Lake.Build.Trace
import Lake.Util.Platform
import Lake.Util.Log

namespace Lake
open Lean System

deriving instance BEq, Hashable for Import

/- Cache for the imported header environment of Lake configuration files. -/
initialize importEnvCache : IO.Ref (HashMap (Array Import) Environment) ← IO.mkRef {}

/-- Like `importModules`, but fetch the resulting import state from the cache if possible. -/
def importModulesUsingCache (imports : Array Import) (opts : Options) (trustLevel : UInt32) : IO Environment := do
  if let some env := (← importEnvCache.get).find? imports then
    return env
  let env ← importModules imports opts trustLevel
  importEnvCache.modify (·.insert imports env)
  return env

/-- Like `Lean.Elab.processHeader`, but using `importEnvCache`. -/
def processHeader (header : Syntax) (opts : Options)
(inputCtx : Parser.InputContext) : StateT MessageLog IO Environment := do
  try
    let imports := Elab.headerToImports header
    importModulesUsingCache imports opts 1024
  catch e =>
    let pos := inputCtx.fileMap.toPosition <| header.getPos?.getD 0
    modify (·.add { fileName := inputCtx.fileName, data := toString e, pos })
    mkEmptyEnvironment

/-- Main module `Name` of a Lake configuration file. -/
def configModuleName : Name := `lakefile

/-- Elaborate `configFile` with the given package directory and options. -/
def elabConfigFile (pkgDir : FilePath) (lakeOpts : NameMap String)
(leanOpts := Options.empty) (configFile := pkgDir / defaultConfigFile) : LogIO Environment := do

  -- Read file and initialize environment
  let input ← IO.FS.readFile configFile
  let inputCtx := Parser.mkInputContext input configFile.toString
  let (header, parserState, messages) ← Parser.parseHeader inputCtx
  let (env, messages) ← processHeader header leanOpts inputCtx messages
  let env := env.setMainModule configModuleName

  -- Configure extensions
  let env := dirExt.setState env pkgDir
  let env := optsExt.setState env lakeOpts

  -- Elaborate File
  let commandState := Elab.Command.mkState env messages leanOpts
  let s ← Elab.IO.processCommands inputCtx parserState commandState

  -- Log messages
  for msg in s.commandState.messages.toList do
    match msg.severity with
    | MessageSeverity.information => logInfo (← msg.toString)
    | MessageSeverity.warning     => logWarning (← msg.toString)
    | MessageSeverity.error       => logError (← msg.toString)

  -- Check result
  if s.commandState.messages.hasErrors then
    error s!"{configFile}: package configuration has errors"
  else
    return s.commandState.env

/--
`Lean.Environment.add` is now private, but exported as `lean_environment_add`.
We call it here via `@[extern]` with a mock implementation.
-/
@[extern "lean_environment_add"]
private opaque addToEnv (env : Environment) (_ : ConstantInfo) : Environment

/--
Import a configuration `.olean` (e.g., `lakefile.olean`).
Auxiliary definition for `importConfigFile`.
-/
def importConfigFileCore (olean : FilePath) (leanOpts : Options) : IO Environment := do
  /-
  We could import the olean together with its imports using one
  `importModules` call, but that would mean we pay for a full
  `finalizeImports` each time, which is linear in the number of imported
  constants and extension entries (in fact, it is paid two more times: when
  marking the `Environment` object as multi-threaded, and when releasing
  it). As most lakefiles use the same set of imports, we instead construct
  an `Environment` for it only once, and then apply the lakefile contents
  on top of it like the elaborator would. Thus the non-shared part of the
  `Environment` is very small.
  -/
  let (mod, _) ← readModuleData olean
  let env ← importModulesUsingCache mod.imports leanOpts 1024
  -- Apply constants (does not go through the kernel, so order is irrelevant)
  let env := mod.constants.foldl addToEnv env
  /-
  Below, we pass module environment extension entries to
  `PersistentEnvExtension.addEntryFn` - but for an extension of
  non-erased type `PersistentEnvExtension α β σ`, the former are of type
  `α` while `addEntryFn` expects `β`s (the type-erased
  `persistentEnvExtensionsRef` ought to be `unsafe` to prevent this from
  simply compiling but it isn't right now). Fortunately, all extensions
  relevant for workspace loading, which we filter for here, have `α = β`;
  entries for any other extensions can safely be ignored.
  -/
  let extDescrs ← persistentEnvExtensionsRef.get
  let extNameIdx ← mkExtNameMap 0
  let env := mod.entries.foldl (init := env) fun env (extName, ents) =>
    if lakeExts.contains extName then
      match extNameIdx.find? extName with
      | some entryIdx => ents.foldl extDescrs[entryIdx]!.addEntry env
      | none => env
    else
      env
  return env
where
  lakeExts :=
    NameSet.empty
    -- Lake Persistent Extensions
    |>.insert ``packageAttr
    |>.insert ``packageDepAttr
    |>.insert ``postUpdateAttr
    |>.insert ``scriptAttr
    |>.insert ``defaultScriptAttr
    |>.insert ``leanLibAttr
    |>.insert ``leanExeAttr
    |>.insert ``externLibAttr
    |>.insert ``targetAttr
    |>.insert ``defaultTargetAttr
    |>.insert ``moduleFacetAttr
    |>.insert ``packageFacetAttr
    |>.insert ``libraryFacetAttr
    -- Docstring Extension (e.g., for scripts)
    |>.insert `Lean.docStringExt
    -- IR Extension (for constant evaluation)
    |>.insert ``IR.declMapExt

instance : ToJson Hash := ⟨(toJson ·.val)⟩
instance : FromJson Hash := ⟨((⟨·⟩) <$> fromJson? ·)⟩

structure ConfigTrace where
  platform : String
  leanHash : String
  configHash : Hash
  options : NameMap String
  deriving ToJson, FromJson

inductive ConfigLock
| olean (h : IO.FS.Handle)
| lean (h : IO.FS.Handle) (lakeOpts : NameMap String)

/--
Import the `.olean` for the configuration file if `reconfigure` is not set and
an up-to-date one exists (i.e., one with matching configuration and on the same
toolchain). Otherwise, elaborate the configuration and save it to the `.olean`.
-/
def importConfigFile (pkgDir lakeDir : FilePath) (lakeOpts : NameMap String)
(leanOpts := Options.empty) (configFile := pkgDir / defaultConfigFile) (reconfigure := true) : LogIO Environment := do
  let some configName := FilePath.mk <$> configFile.fileName
    | error "invalid configuration file name"
  let olean := lakeDir / configName.withExtension "olean"
  let traceFile := lakeDir / configName.withExtension "olean.trace"
  /-
  Remark: To prevent race conditions between the `.olean` and its trace file
  (i.e., one process writes a new configuration while another reads an old one
  and vice versa), Lake takes opens a single handle on the trace file and
  acquires a lock on it. The lock is held while the lock data is read and
  the olean is either imported or a new one is written with new lock data.
  -/
  let configHash ← computeTextFileHash configFile
  let configLock : ConfigLock ← id do
    let validateTrace h : IO ConfigLock := id do
      if reconfigure then
        h.lock; return .lean h lakeOpts
      h.lock (exclusive := false)
      let contents ← h.readToEnd; h.rewind
      let .ok (trace : ConfigTrace) := Json.parse contents >>= fromJson?
        | error "compiled configuration is invalid; run with `-R` to reconfigure"
      let upToDate := trace.platform = platformDescriptor ∧
        trace.leanHash = Lean.githash ∧ trace.configHash = configHash
      if upToDate then
        return .olean h
      else
        -- Upgrade to an exclusive lock
        let lockFile := lakeDir / configName.withExtension "olean.lock"
        let hLock ← IO.FS.Handle.mk lockFile .write
        hLock.lock; h.unlock; h.lock; hLock.unlock
        return .lean h trace.options
    if (← traceFile.pathExists) then
      validateTrace <| ← IO.FS.Handle.mk traceFile .readWrite
    else
      IO.FS.createDirAll lakeDir
      match (← IO.FS.Handle.mk traceFile .writeNew |>.toBaseIO) with
      | .ok h =>
        h.lock; return .lean h lakeOpts
      | .error (.alreadyExists ..) =>
        validateTrace <| ← IO.FS.Handle.mk traceFile .readWrite
      | .error e => error e.toString
  match configLock with
  | .olean h =>
    let env ← importConfigFileCore olean leanOpts
    h.unlock
    return env
  | .lean h lakeOpts =>
    let env ← elabConfigFile pkgDir lakeOpts leanOpts configFile
    Lean.writeModule env olean
    h.putStrLn <| Json.pretty <| toJson
      {platform := platformDescriptor, leanHash := Lean.githash,
        configHash, options := lakeOpts : ConfigTrace}
    h.truncate
    h.unlock
    return env
