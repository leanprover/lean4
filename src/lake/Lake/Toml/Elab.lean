/-
Copyright (c) 2024 Mac Malone. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Mac Malone
-/
prelude
import Lake.Toml.Elab.Expression

/-!
# TOML Elaboration

Elaborates TOML syntax into Lean data types.
At top-level, elaborates a whole TOML file into a `Toml.Table`.
-/
