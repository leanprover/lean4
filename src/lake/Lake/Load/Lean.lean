/-
Copyright (c) 2024 Mac Malone. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Mac Malone
-/
import Lake.Load.Lean.Elab
import Lake.Load.Lean.Eval

/-! # Lean Loader

This module contains the main definitions to load a package from a
Lake configuration file written in Lake's Lean DSL.
-/

open Lean

namespace Lake

/--
Elaborate a Lean configuration file into a `Package`.
The resulting package does not yet include any dependencies.
-/
def loadLeanConfig (cfg : LoadConfig)
: LogIO (Package × Environment) := do
  let configEnv ← importConfigFile cfg
  let pkgConfig ← IO.ofExcept <| PackageConfig.loadFromEnv configEnv cfg.leanOpts
  let pkg : Package := {
    dir := cfg.pkgDir
    relDir := cfg.relPkgDir
    config := pkgConfig
    relConfigFile := cfg.relConfigFile
    remoteUrl? := cfg.remoteUrl?
  }
  return (← pkg.loadFromEnv configEnv cfg.leanOpts, configEnv)
