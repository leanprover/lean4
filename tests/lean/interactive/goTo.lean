import Lean.Elab

structure Bar where

structure Foo where
  foo₁ : Nat
  foo₂ : Nat
  bar  : Bar

def mkFoo₁ : Foo := {
--v textDocument/definition
  foo₁ := 1
--v textDocument/declaration
  foo₂ := 2
--v textDocument/typeDefinition
  bar := ⟨⟩
}

structure HandWrittenStruct where
  n : Nat

-- def HandWrittenStruct.n := fun | mk n => n

          --v textDocument/definition
def hws : HandWrittenStruct := {
--v textDocument/definition
  n := 3
}

            --v textDocument/declaration
def mkFoo₂ := mkFoo₁

syntax (name := elabTest) "test" : term

@[termElab elabTest] def elabElabTest : Lean.Elab.Term.TermElab := fun _ _ => do
  let stx ← `(2)
  Lean.Elab.Term.elabTerm stx none

     --v textDocument/declaration
#check test
     --^ textDocument/definition

def Baz (α : Type) := α

#check fun (b : Baz Nat) => b
                          --^ textDocument/typeDefinition

example : Nat :=
  let a := 1
--v textDocument/definition
  a + b
    --^ textDocument/definition
where
  b := 2

macro_rules | `(test) => `(3)
#check test
     --^ textDocument/definition
