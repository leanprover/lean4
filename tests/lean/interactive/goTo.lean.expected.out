{"textDocument": {"uri": "file://goTo.lean"},
 "position": {"line": 11, "character": 2}}
[{"targetUri": "file://goTo.lean",
  "targetSelectionRange":
  {"start": {"line": 5, "character": 2}, "end": {"line": 5, "character": 6}},
  "targetRange":
  {"start": {"line": 5, "character": 2}, "end": {"line": 5, "character": 6}},
  "originSelectionRange":
  {"start": {"line": 11, "character": 2}, "end": {"line": 11, "character": 6}}}]
{"textDocument": {"uri": "file://goTo.lean"},
 "position": {"line": 13, "character": 2}}
[{"targetUri": "file://goTo.lean",
  "targetSelectionRange":
  {"start": {"line": 6, "character": 2}, "end": {"line": 6, "character": 6}},
  "targetRange":
  {"start": {"line": 6, "character": 2}, "end": {"line": 6, "character": 6}},
  "originSelectionRange":
  {"start": {"line": 13, "character": 2}, "end": {"line": 13, "character": 6}}}]
{"textDocument": {"uri": "file://goTo.lean"},
 "position": {"line": 15, "character": 2}}
[{"targetUri": "file://goTo.lean",
  "targetSelectionRange":
  {"start": {"line": 2, "character": 10}, "end": {"line": 2, "character": 13}},
  "targetRange":
  {"start": {"line": 2, "character": 0}, "end": {"line": 2, "character": 19}},
  "originSelectionRange":
  {"start": {"line": 15, "character": 2}, "end": {"line": 15, "character": 5}}}]
{"textDocument": {"uri": "file://goTo.lean"},
 "position": {"line": 24, "character": 12}}
[{"targetUri": "file://goTo.lean",
  "targetSelectionRange":
  {"start": {"line": 18, "character": 10},
   "end": {"line": 18, "character": 27}},
  "targetRange":
  {"start": {"line": 18, "character": 0}, "end": {"line": 19, "character": 9}},
  "originSelectionRange":
  {"start": {"line": 24, "character": 10},
   "end": {"line": 24, "character": 27}}}]
{"textDocument": {"uri": "file://goTo.lean"},
 "position": {"line": 26, "character": 2}}
[{"targetUri": "file://goTo.lean",
  "targetSelectionRange":
  {"start": {"line": 19, "character": 2}, "end": {"line": 19, "character": 3}},
  "targetRange":
  {"start": {"line": 19, "character": 2}, "end": {"line": 19, "character": 3}},
  "originSelectionRange":
  {"start": {"line": 26, "character": 2}, "end": {"line": 26, "character": 3}}}]
{"textDocument": {"uri": "file://goTo.lean"},
 "position": {"line": 30, "character": 14}}
[{"targetUri": "file://goTo.lean",
  "targetSelectionRange":
  {"start": {"line": 9, "character": 4}, "end": {"line": 9, "character": 10}},
  "targetRange":
  {"start": {"line": 9, "character": 0}, "end": {"line": 16, "character": 1}},
  "originSelectionRange":
  {"start": {"line": 30, "character": 14},
   "end": {"line": 30, "character": 20}}}]
{"textDocument": {"uri": "file://goTo.lean"},
 "position": {"line": 39, "character": 7}}
[{"targetUri": "file://goTo.lean",
  "targetSelectionRange":
  {"start": {"line": 32, "character": 0}, "end": {"line": 32, "character": 39}},
  "targetRange":
  {"start": {"line": 32, "character": 0}, "end": {"line": 32, "character": 32}},
  "originSelectionRange":
  {"start": {"line": 39, "character": 7},
   "end": {"line": 39, "character": 11}}}]
{"textDocument": {"uri": "file://goTo.lean"},
 "position": {"line": 39, "character": 7}}
[{"targetUri": "file://goTo.lean",
  "targetSelectionRange":
  {"start": {"line": 34, "character": 25},
   "end": {"line": 34, "character": 37}},
  "targetRange":
  {"start": {"line": 34, "character": 21},
   "end": {"line": 36, "character": 34}},
  "originSelectionRange":
  {"start": {"line": 39, "character": 7},
   "end": {"line": 39, "character": 11}}}]
{"textDocument": {"uri": "file://goTo.lean"},
 "position": {"line": 44, "character": 28}}
[{"targetUri": "file://goTo.lean",
  "targetSelectionRange":
  {"start": {"line": 42, "character": 4}, "end": {"line": 42, "character": 7}},
  "targetRange":
  {"start": {"line": 42, "character": 0}, "end": {"line": 42, "character": 23}},
  "originSelectionRange":
  {"start": {"line": 44, "character": 28},
   "end": {"line": 44, "character": 29}}}]
{"textDocument": {"uri": "file://goTo.lean"},
 "position": {"line": 50, "character": 2}}
[{"targetUri": "file://goTo.lean",
  "targetSelectionRange":
  {"start": {"line": 48, "character": 6}, "end": {"line": 48, "character": 7}},
  "targetRange":
  {"start": {"line": 48, "character": 6}, "end": {"line": 48, "character": 7}},
  "originSelectionRange":
  {"start": {"line": 50, "character": 2}, "end": {"line": 50, "character": 3}}}]
{"textDocument": {"uri": "file://goTo.lean"},
 "position": {"line": 50, "character": 6}}
[{"targetUri": "file://goTo.lean",
  "targetSelectionRange":
  {"start": {"line": 53, "character": 2}, "end": {"line": 53, "character": 3}},
  "targetRange":
  {"start": {"line": 53, "character": 2}, "end": {"line": 53, "character": 3}},
  "originSelectionRange":
  {"start": {"line": 50, "character": 6}, "end": {"line": 50, "character": 7}}}]
{"textDocument": {"uri": "file://goTo.lean"},
 "position": {"line": 56, "character": 7}}
[{"targetUri": "file://goTo.lean",
  "targetSelectionRange":
  {"start": {"line": 55, "character": 0}, "end": {"line": 55, "character": 11}},
  "targetRange":
  {"start": {"line": 55, "character": 0}, "end": {"line": 55, "character": 29}},
  "originSelectionRange":
  {"start": {"line": 56, "character": 7},
   "end": {"line": 56, "character": 11}}}]
